CloudMigration: Removes snapshot and resources when deleting a session

leandro-deveikis opened the pull request:
Fixes https://github.com/grafana/grafana-operator-experience-squad/issues/908

Removes all associated data when deleting a session:
- Snapshot resources (DB)
- Snapshots (DB)
- Snapshot files from the file system

leandro-deveikis committed:
CloudMigration: Removes snapshot and resources when deleting a session

leandro-deveikis committed:
fix comment

PoorlyDefinedBehaviour commented on pkg/services/cloudmigration/cloudmigrationimpl/cloudmigration.go:
```diff
@@ -464,8 +464,24 @@ func (s *Service) GetMigrationRunList(ctx context.Context, migUID string) (*clou
-	c, err := s.store.DeleteMigrationSessionByUID(ctx, uid)
+func (s *Service) DeleteSession(ctx context.Context, sessionUID string) (*cloudmigration.CloudMigrationSession, error) {
+	// first we try to delete all the associated information to the session
+	snapshots, err := s.store.GetMigrationStatusList(ctx, sessionUID)
```
Can we use a transaction to avoid deleting only part of what needs to be deleted?

leandro-deveikis replied:
done! I moved them to the sql store and run everything within a transaction

PoorlyDefinedBehaviour commented on pkg/services/cloudmigration/cloudmigrationimpl/cloudmigration.go:
```diff
@@ -464,8 +464,24 @@ func (s *Service) GetMigrationRunList(ctx context.Context, migUID string) (*clou
 
-func (s *Service) DeleteSession(ctx context.Context, uid string) (*cloudmigration.CloudMigrationSession, error) {
-	c, err := s.store.DeleteMigrationSessionByUID(ctx, uid)
+func (s *Service) DeleteSession(ctx context.Context, sessionUID string) (*cloudmigration.CloudMigrationSession, error) {
```
Can we delete the snapshot files as well since they won't be used anymore and are occupying disk space?

leandro-deveikis replied:
done!

mmandrus commented on pkg/services/cloudmigration/cloudmigrationimpl/cloudmigration_test.go:
```diff
@@ -115,6 +115,12 @@ func Test_CreateGetRunMigrationsAndRuns(t *testing.T) {
+
+	// after deleting the session, the snapshots and resources should not exist anymore.
+	// we check the snapshot for now
+	listRunResp2, err := s.GetMigrationRunList(context.Background(), createResp.UID)
```
This unit test also tests the to-be-removed workflow which is a little weird. What do you think about converting this unit test to use the new APIs? The equivalent would be seeding the store with some snapshot results.

leandro-deveikis replied:
I think I would prefer to do this in a new PR, when we clean the old func (which I am not 100% sure which ones are) and we can merge this one

mmandrus replied:
I'm fine leaving unit test cleanup for later

leandro-deveikis committed:
Removing snapshot files

leandro-deveikis committed:
adding deprecation comment

leandro-deveikis committed:
Fix using new func

leandro-deveikis committed:
move delete to a new function

leandro-deveikis committed:
remove test

leandro-deveikis committed:
Add comment

leandro-deveikis committed:
Move deletes to sqlstore and run them in a transaction

leandro-deveikis commented on pkg/services/cloudmigration/cloudmigrationimpl/cloudmigration.go:
```diff
@@ -464,8 +464,24 @@ func (s *Service) GetMigrationRunList(ctx context.Context, migUID string) (*clou
-	c, err := s.store.DeleteMigrationSessionByUID(ctx, uid)
+func (s *Service) DeleteSession(ctx context.Context, sessionUID string) (*cloudmigration.CloudMigrationSession, error) {
+	// first we try to delete all the associated information to the session
+	snapshots, err := s.store.GetMigrationStatusList(ctx, sessionUID)
```
done!

mmandrus commented on pkg/services/cloudmigration/cloudmigrationimpl/cloudmigration.go:
```diff
@@ -96,9 +96,11 @@ func ProvideService(
 	s := &Service{
-		store:            &sqlStore{db: db, secretsStore: secretsStore, secretsService: secretsService},
-		log:              log.New(LogPrefix),
+		store:            &sqlStore{db: db, secretsStore: secretsStore, secretsService: secretsService, log: logger},
```
IMO, this subcomponent should initialize its own logger

mmandrus replied:
Once the delete snapshot logic is moved up a level, the log won't be needed anyway

mmandrus commented on pkg/services/cloudmigration/cloudmigrationimpl/cloudmigration.go:
```diff
@@ -464,14 +466,13 @@ func (s *Service) GetMigrationRunList(ctx context.Context, migUID string) (*clou
-func (s *Service) DeleteSession(ctx context.Context, uid string) (*cloudmigration.CloudMigrationSession, error) {
-	c, err := s.store.DeleteMigrationSessionByUID(ctx, uid)
+func (s *Service) DeleteSession(ctx context.Context, sessionUID string) (*cloudmigration.CloudMigrationSession, error) {
+	c, err := s.store.DeleteMigrationSessionWithRelatedElements(ctx, sessionUID)
```
I think we should just add your functionality from your new function into the existing `DeleteMigrationSessionByUID` function, and just add documentation noting that the function deletes related elements.

mmandrus commented on pkg/services/cloudmigration/cloudmigrationimpl/xorm_store.go:
```diff
@@ -259,11 +271,14 @@ func (ss *sqlStore) GetSnapshotByUID(ctx context.Context, sessionUid, uid string
 }
 
 // GetSnapshotList returns snapshots without resources included. Use GetSnapshotByUID to get individual snapshot results.
+// passing a limit of -1 will return all the elements regardless of the page
```
I think we should add a constant in here so that we can pass in `cloudmigrationimpl.AllSnapshots` instead

mmandrus commented on pkg/services/cloudmigration/cloudmigrationimpl/xorm_store.go:
```diff
@@ -433,3 +448,45 @@ func (ss *sqlStore) decryptToken(ctx context.Context, cm *cloudmigration.CloudMi
 }
+
+func (ss *sqlStore) DeleteMigrationSessionWithRelatedElements(ctx context.Context, sessionUID string) (*cloudmigration.CloudMigrationSession, error) {
+	// first we try to delete all the associated information to the session
```
All of the database transactions need to be wrapped in an `InTransaction` block, e.g. [here](https://github.com/grafana/grafana/blob/eb72c3f8d3e83e04620ec524de3a8331a462ac73/pkg/services/cloudmigration/cloudmigrationimpl/xorm_store.go#L317) to ensure atomicity

mmandrus commented on pkg/services/cloudmigration/cloudmigrationimpl/xorm_store.go:
```diff
@@ -433,3 +448,45 @@ func (ss *sqlStore) decryptToken(ctx context.Context, cm *cloudmigration.CloudMi
+			return nil, fmt.Errorf("deleting snapshot from db: %w", err)
+		}
+
+		err = deleteLocalFiles(snapshot)
```
I don't think the store should be responsible for deleting the local files, that should be done by the caller after the fact. It might be slightly weird, but my workaround would be to just add the list of localfiles to this method's return values.

mmandrus reviewed (changes requested):
Looking a lot cleaner. IMO we can just use the existing delete session function instead of making a new one with a long name. Some other minor changes requested as well. Nearly there!

leandro-deveikis committed:
- Moving delete file to cloudmigration service
- Removing DeleteMigrationSessionWithRelatedElements and moving logic back to DeleteMigrationSessionByUID
- Implementing transaction for db changes

leandro-deveikis committed:
Add GetAllSnapshots const

leandro-deveikis committed:
add comment

leandro-deveikis committed:
add call to event when error

diegommm commented:
Hi! Greetings from the Release Guild. This is a friendly reminder that the Steady [11.2.x Release Branch](https://github.com/grafana/grafana/tree/v11.2.x) was cut this week, and that it includes the changes committed up to https://github.com/grafana/grafana/commit/b4126d3bce3627e9460e2db0e870c8d984be8e84. This PR was merged after it, so we wanted to let you know that you can add the `backport v11.2.x` label to to backport it and make it part of the next 11.2.x Patch Release, just in case you may have missed the news.

Thank you!

yuri-tceretian commented:
Greetings from Release Guild. This PR has been successfully included in the Grafana 11.2 release branch. A backport PR was also created to ensure the v11.2.x branch is correctly synchronized with the release.
